### PR TITLE
Fix off-by-one Default Output token visualization in Gemma tracer steering (#204)

### LINKED ISSUE #204: Off-by-1 token visualization in Gemma-2-2B tracer graph steering
## Bug Description
In Gemma-2-2B's Circuit Tracer's steering interface, hovering over a Default Output token after steering shows the probabilities for the current token instead of the next one.

Note that Qwen3-4B's steering behaves as expected.

## How To Reproduce Bug
1. Open the default Circuit Tracer for Gemma-2-2B
2. Click "Steer" in Subgraph pane
3. Steer some feature
4. Hover over a token under Default Output. 

Unexpected: Instead of showing probabilities for the following token prediction, they are shown for the current token.

## Screenshots

Image

## Additional Context

This is apparently a similar issue to the one previously fixed in 91f875d5ac5c7257d1db67f90b5a8d20943ddec3: A duplicate prepended BOS token due to re-tokenization. Claude's proposed fix was to replace https://github.com/hijohnnylin/neuronpedia/blob/0931811f5bed602142cab0f95b2e1e5e021a09ff/apps/graph/runpod-gemma-2-2b/rp_handler.py#L342 with code similar to the one in that patch. (I did not make the fix myself because I'm still, unfortunately, unable to get the Circuit Tracer running locally to test it).

### PR DESCRIPTION
**Pull request by Alexander230**
Fixes #204.

## The bug (Gemma)

In the Gemma-2-2B circuit-tracer steering UI, hovering a **Default Output** token shows the top logits for the *current* token instead of the *next* one.

Root cause is in `apps/graph/runpod-gemma-2-2b/rp_handler.py`. The default-output logits are computed by re-running the model on the **decoded generation string**:

```python
default_logits = model(default_generation)   # default_generation = "".join(decoded tokens)
```

`default_tokenized` already starts with Gemma's `<bos>` (id 2). Decoding it to text and passing that string back through `model(...)` re-tokenizes it, and Gemma's tokenizer prepends `<bos>` **again** — so `default_logits` is computed over a sequence with a **duplicate BOS**, and every logit position is shifted by one relative to `default_tokenized_str_tokens`. The per-token hover then reads the logits for position *i-1* while labeling them position *i*.

This is the same class of bug fixed for `neuronpedia_graph/server.py` in 91f875d; it was never ported to this RunPod handler.

## The fix

Pass the token IDs directly instead of re-tokenizing a decoded string, so no duplicate BOS is introduced:

```python
default_logits = model(default_tokenized.unsqueeze(0))
if default_logits.dim() == 2:
    default_logits = default_logits.unsqueeze(0)
```

## Verified on the real model

Ran circuit-tracer (the rev this handler pins, `e4a3c5af`) on `google/gemma-2-2b`. With greedy decoding the top-1 prediction at position *i* must equal the actual next token *i+1*:

- Re-tokenizing the decoded generation produces length 15 vs. the real 14 → **duplicate BOS confirmed**.
- Default-output hover, next-token match: **before 0/8, after 8/8**. Before the fix the hover shows the current token at every position (exactly the reported symptom); after, it shows the correct next token.

## Scope notes

**The steered-output loop is intentionally left unchanged.** `server.py`'s later fix also offsets the *steered* logits index (`gen_idx = i - (sequence_length - 1)`), but that is only correct for the circuit-tracer version `server.py` pins (`tag v0.3.1`). This handler pins `rev e4a3c5af`, where `feature_intervention_generate` returns **full-sequence** logits — so the existing `steered_logits[:, : i + 1, :]` indexing is already correct here. Verified on the real model (empty and real interventions, greedy oracle): the existing indexing scores 9/9, the offset scores 0–1/9. Applying the offset would *break* the steered hover, so this PR touches only the default-output path.

**The `runpod-qwen3-4b` handler does not need this fix.** It has identical code, but Qwen3's tokenizer has no BOS token (`bos_token = None`), so decoding and re-tokenizing the generation introduces no duplicate token and the positions are already aligned — its default-output hover is correct as-is (consistent with the issue reporter noting Qwen behaves as expected). It pins the same circuit-tracer rev, so its steered path is likewise already correct. Left untouched to keep this change minimal.

### CODE REVIEW

**Comment by hijohnnylin** [comment]
Hm this may have been resolved by a recent circuit-tracer library update (we just updated to 0.5.1) - can you double check?
image

**Comment by Alexander230** [comment]
I've tested on the real gemma-2-2b using the stack the graph app now locks (circuit-tracer tag v0.5.1, transformer-lens 3.2.1, transformers 4.57.3): the library update alone doesn't fix this for the runpod handler. Unpatched model(default_generation) still re-tokenizes with a duplicate BOS (15 vs 14 tokens), and the Default Output hover matches the true next token 0/8 — identical to the current pin. With this PR's change: 8/8. v0.5.1's ensure_tokenized() does dedup a leading BOS, but the plain model(string) forward path never routes through it.

It can look resolved on your side because the graph server has had the same pass-token-ids fix since 91f875d (March), and that's the app the 0.5.1 bump (79350ad) updated — the runpod handlers still pin safety-research/circuit-tracer@e4a3c5af, which is where this PR applies.

A change to this PR is needed if you plan to bump the handlers to 0.5.1: feature_intervention_generate now returns 2-D generation-only logits (changed in a1ca600), so the handler's steered loop steered_logits[:, : i + 1, :] raises IndexError as-is, and with just a 2-D guard the absolute-i indexing shows wrong tokens (0/9 vs 9/9 for server.py's gen_idx offset). This patch matches the handlers' current pin, and with 0.5.1 they need to be changed to gen_idx offset + 2-D guard.

**Comment by hijohnnylin** [comment]
Ah thanks for clarifying! Sorry I missed that this PR was specifically about the runpod routes.
We did switch away from using the runpod handlers recently on prod and are moving away from vendor-specific code in general, which explains why I was unable to reproduce the issue in prod. I'll merge your fix here in case anyone is using it still.
